internal/dag/gatewayapi_processor_test.go

```diff
@@ -17,175 +17,190 @@ import (
+			hostnames:    []gatewayapi_v1alpha1.Hostname{},
+			want:         []string(nil),
+			wantError: []error{
+				fmt.Errorf("invalid hostname \"#projectcontour.io\": [a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"),
+			},
+		},
+		"invalid listener wildcard hostname": {
+			listenerHost: listenerHostname("*.*.projectcontour.io"),
+			hostnames:    []gatewayapi_v1alpha1.Hostname{},
+			want:         []string(nil),
+			wantError: []error{
+				fmt.Errorf("invalid hostname \"*.*.projectcontour.io\": [a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\\*\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"),
+			},
+		},
+		"listener host & hostnames host do not exactly match": {
+			listenerHost: listenerHostname("listener.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want:      nil,
+			wantError: []error{fmt.Errorf("gateway hostname \"listener.projectcontour.io\" does not match hostnames hostname \"http.projectcontour.io\"")},
+		},
+		"listener host & hostnames host exactly match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want:      []string{"http.projectcontour.io"},
+			wantError: nil,
+		},
+		"listener host & hostnames host match wildcard host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want:      []string{"http.projectcontour.io"},
+			wantError: nil,
+		},
+		"listener host & hostnames host do not match wildcard host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.example.com",
 			},
+			want:      nil,
+			wantError: []error{fmt.Errorf("gateway hostname \"*.projectcontour.io\" does not match hostnames hostname \"http.example.com\"")},
+		},
+		"listener host & wildcard hostnames host match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+			},
+			want:      []string{"*.projectcontour.io"},
+			wantError: nil,
+		},
+		"listener host wildcard & wildcard hostnames host match": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"*.projectcontour.io",
+			},
+			want:      []string{"*.projectcontour.io"},
+			wantError: nil,
+		},
+		"listener host & wildcard all hostnames host match": {
```

Comment:
similar to the above, seems like this is a more specific gateway hostname than the route hostname

Reply:
I guess this test would be the same as (https://github.com/projectcontour/contour/blob/d0ed425ce3a3d6950e989e3ae58963fea5e2bfae/internal/dag/gatewayapi_processor_test.go#L183) which essentially has no hostname defined on the Route, but the Gateway does, so it should end up only matching the Gateway.